File: android/app/src/main/java/com/noop/analytics/SleepPlanner.kt
package com.noop.analytics

import kotlin.math.abs
import kotlin.math.ceil

enum class SleepGoalMode(val persistedValue: String) {
    TARGET("target"),
    BALANCE("balance"),
    EXTRA_OPPORTUNITY("extraOpportunity");

    companion object {
        fun fromPersisted(value: String?): SleepGoalMode =
            entries.firstOrNull { it.persistedValue == value } ?: BALANCE
    }
}

/**
 * One deterministic contract for tonight's sleep opportunity.
 *
 * This is deliberately not a clinical prescription. It combines the user's wake
 * time and explicit sleep target with a short wind-down buffer and, when enough
 * history exists, a bounded contribution from the recent arithmetic sleep balance.
 * Mirrors StrandAnalytics/SleepPlanner.swift value-for-value.
 */
data class SleepPlan(
    val wakeMinute: Int,
    val bedtimeMinute: Int,
    val windDownMinute: Int,
    val baseSleepMinutes: Int,
    val goalMode: SleepGoalMode,
    val recoveryMinutes: Int,
    val sleepOpportunityMinutes: Int,
    val debtBalanceMinutes: Double?,
    val historyNights: Int,
    val confidence: ScoreConfidence,
)

object SleepPlanner {
    const val MINIMUM_SLEEP_MINUTES = 5 * 60
    const val MAXIMUM_SLEEP_MINUTES = 11 * 60
    const val MAXIMUM_RECOVERY_MINUTES = 60
    const val RECOVERY_STEP_MINUTES = 15
    const val MINIMUM_DEBT_NIGHTS = 3
    const val SOLID_HISTORY_NIGHTS = 7

    fun plan(
        wakeMinute: Int,
        sleepTargetMinutes: Int,
        windDownLeadMinutes: Int,
        debtBalanceMinutes: Double?,
        historyNights: Int,
        goalMode: SleepGoalMode = SleepGoalMode.BALANCE,
    ): SleepPlan {
        val wake = wrappedMinute(wakeMinute)
        val base = sleepTargetMinutes.coerceIn(MINIMUM_SLEEP_MINUTES, MAXIMUM_SLEEP_MINUTES)
        val lead = windDownLeadMinutes.coerceIn(0, 120)
        val nights = historyNights.coerceAtLeast(0)
        val balanceRecovery = recoveryMinutes(debtBalanceMinutes, nights)
        val recovery = when (goalMode) {
            SleepGoalMode.TARGET -> 0
            SleepGoalMode.BALANCE -> balanceRecovery
            SleepGoalMode.EXTRA_OPPORTUNITY ->
                balanceRecovery.coerceAtLeast(30).coerceAtMost(MAXIMUM_RECOVERY_MINUTES)
        }
        val opportunity = base + recovery
        val bedtime = wrappedMinute(wake - opportunity)
        val windDown = wrappedMinute(bedtime - lead)
        val confidence = when {
            nights < MINIMUM_DEBT_NIGHTS -> ScoreConfidence.CALIBRATING
            nights < SOLID_HISTORY_NIGHTS -> ScoreConfidence.BUILDING
            else -> ScoreConfidence.SOLID
        }
        return SleepPlan(
            wakeMinute = wake,
            bedtimeMinute = bedtime,
            windDownMinute = windDown,
            baseSleepMinutes = base,
            goalMode = goalMode,
            recoveryMinutes = recovery,
            sleepOpportunityMinutes = opportunity,
            debtBalanceMinutes = debtBalanceMinutes,
            historyNights = nights,
            confidence = confidence,
        )
    }

    /**
     * Spread recent debt over the nights that produced it, round upward to an
     * actionable 15-minute increment, and cap tonight's addition at one hour.
     * A surplus, the ±30-minute deadband, or fewer than three nights adds nothing.
     */
    fun recoveryMinutes(debtBalanceMinutes: Double?, historyNights: Int): Int {
        val balance = debtBalanceMinutes ?: return 0
        if (historyNights < MINIMUM_DEBT_NIGHTS || balance >= -SleepDebt.ON_TARGET_BAND_MIN) return 0
        val averageDeficit = abs(balance) / historyNights.toDouble()
        val stepped = ceil(averageDeficit / RECOVERY_STEP_MINUTES).toInt() * RECOVERY_STEP_MINUTES
        return stepped.coerceIn(RECOVERY_STEP_MINUTES, MAXIMUM_RECOVERY_MINUTES)
    }

    fun wrappedMinute(minute: Int): Int {
        val day = 24 * 60
        return ((minute % day) + day) % day
    }

    /**
     * Signed shortest difference between tonight's planned midpoint and the midpoint observed in
     * prior main sleeps. This is behavioral timing context, not a chronotype or medical inference.
     */
    fun observedTimingShiftMinutes(plan: SleepPlan, habitualMidsleepSeconds: Long?): Int? {
        val seconds = habitualMidsleepSeconds ?: return null
        val observed = wrappedMinute((seconds / 60L).toInt())
        val planned = wrappedMinute(plan.bedtimeMinute + plan.sleepOpportunityMinutes / 2)
        var delta = planned - observed
        if (delta > 12 * 60) delta -= 24 * 60
        if (delta < -(12 * 60)) delta += 24 * 60
        return delta
    }
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/SleepPlanner.swift
import Foundation

public enum SleepGoalMode: String, CaseIterable, Codable, Sendable {
    /// Keep the user-entered target fixed; recent balance is shown but does not change tonight.
    case target
    /// Add only the bounded amount supported by the recent arithmetic sleep-balance ledger.
    case balance
    /// Reserve at least 30 extra minutes of opportunity, still capped at one hour total.
    case extraOpportunity
}

// SleepPlanner.swift — one deterministic contract for tonight's sleep opportunity.
//
// This planner deliberately does less than a clinical sleep prescription. It combines:
//   • the wake time the user chose,
//   • the user's explicit sleep target,
//   • a short wind-down buffer, and
//   • a bounded contribution from the recent arithmetic sleep-balance ledger.
//
// It never reduces the user's target because of a recent surplus, never tries to repay
// an entire multi-night deficit in one night, and never invents a personalized plan when
// fewer than three usable nights exist. The Kotlin mirror must stay value-for-value.

public struct SleepPlan: Equatable, Sendable {
    /// User-selected wake minute in [0, 1439].
    public let wakeMinute: Int
    /// Suggested lights-out minute in [0, 1439].
    public let bedtimeMinute: Int
    /// Suggested start-winding-down minute in [0, 1439].
    public let windDownMinute: Int
    /// User-controlled baseline target, clamped to 5–11 hours.
    public let baseSleepMinutes: Int
    /// User-selected planning behavior.
    public let goalMode: SleepGoalMode
    /// Extra opportunity selected by the goal mode, always 0–60 minutes.
    public let recoveryMinutes: Int
    /// Total planned sleep opportunity: base + recovery.
    public let sleepOpportunityMinutes: Int
    /// Recent arithmetic balance passed by the caller. Negative means debt.
    public let debtBalanceMinutes: Double?
    /// Number of usable nights behind that balance.
    public let historyNights: Int
    /// Certainty of the planning basis, not certainty that sleep will occur as planned.
    public let confidence: ScoreConfidence

    public init(
        wakeMinute: Int,
        bedtimeMinute: Int,
        windDownMinute: Int,
        baseSleepMinutes: Int,
        goalMode: SleepGoalMode,
        recoveryMinutes: Int,
        sleepOpportunityMinutes: Int,
        debtBalanceMinutes: Double?,
        historyNights: Int,
        confidence: ScoreConfidence
    ) {
        self.wakeMinute = wakeMinute
        self.bedtimeMinute = bedtimeMinute
        self.windDownMinute = windDownMinute
        self.baseSleepMinutes = baseSleepMinutes
        self.goalMode = goalMode
        self.recoveryMinutes = recoveryMinutes
        self.sleepOpportunityMinutes = sleepOpportunityMinutes
        self.debtBalanceMinutes = debtBalanceMinutes
        self.historyNights = historyNights
        self.confidence = confidence
    }
}

public enum SleepPlanner {
    public static let minimumSleepMinutes = 5 * 60
    public static let maximumSleepMinutes = 11 * 60
    public static let maximumRecoveryMinutes = 60
    public static let recoveryStepMinutes = 15
    public static let minimumDebtNights = 3
    public static let solidHistoryNights = 7

    /// Build tonight's plan. `debtBalanceMinutes` should come from `SleepDebt.ledger`
    /// using the same explicit sleep target. Negative balance may add a small recovery
    /// buffer; a positive balance never subtracts from the target.
    public static func plan(
        wakeMinute: Int,
        sleepTargetMinutes: Int,
        windDownLeadMinutes: Int,
        debtBalanceMinutes: Double?,
        historyNights: Int,
        goalMode: SleepGoalMode = .balance
    ) -> SleepPlan {
        let wake = wrappedMinute(wakeMinute)
        let base = min(max(sleepTargetMinutes, minimumSleepMinutes), maximumSleepMinutes)
        let lead = min(max(windDownLeadMinutes, 0), 120)
        let nights = max(historyNights, 0)
        let balanceRecovery = recoveryMinutes(
            debtBalanceMinutes: debtBalanceMinutes,
            historyNights: nights
        )
        let recovery: Int = {
            switch goalMode {
            case .target:
                return 0
            case .balance:
                return balanceRecovery
            case .extraOpportunity:
                return min(max(balanceRecovery, 30), maximumRecoveryMinutes)
            }
        }()
        let opportunity = base + recovery
        let bedtime = wrappedMinute(wake - opportunity)
        let windDown = wrappedMinute(bedtime - lead)
        let confidence: ScoreConfidence = {
            if nights < minimumDebtNights { return .calibrating }
            if nights < solidHistoryNights { return .building }
            return .solid
        }()

        return SleepPlan(
            wakeMinute: wake,
            bedtimeMinute: bedtime,
            windDownMinute: windDown,
            baseSleepMinutes: base,
            goalMode: goalMode,
            recoveryMinutes: recovery,
            sleepOpportunityMinutes: opportunity,
            debtBalanceMinutes: debtBalanceMinutes,
            historyNights: nights,
            confidence: confidence
        )
    }

    /// Spread recent debt over the nights that produced it, round upward to an
    /// actionable 15-minute increment, and cap tonight's addition at one hour.
    ///
    /// A balance inside SleepDebt's ±30-minute deadband, a surplus, or fewer than
    /// three usable nights yields no adjustment.
    public static func recoveryMinutes(
        debtBalanceMinutes: Double?,
        historyNights: Int
    ) -> Int {
        guard historyNights >= minimumDebtNights,
              let balance = debtBalanceMinutes,
              balance < -SleepDebt.onTargetBandMin else { return 0 }

        let averageDeficit = abs(balance) / Double(historyNights)
        let stepped = Int(ceil(averageDeficit / Double(recoveryStepMinutes))) * recoveryStepMinutes
        return min(max(stepped, recoveryStepMinutes), maximumRecoveryMinutes)
    }

    public static func wrappedMinute(_ minute: Int) -> Int {
        let day = 24 * 60
        return ((minute % day) + day) % day
    }

    /// Difference between tonight's planned midpoint and the midpoint observed in prior main sleeps.
    /// Positive values mean later; negative values mean earlier. This is behavioral timing context,
    /// not a biological chronotype or medical inference.
    public static func observedTimingShiftMinutes(
        plan: SleepPlan,
        habitualMidsleepSeconds: Int?
    ) -> Int? {
        guard let seconds = habitualMidsleepSeconds else { return nil }
        let observed = wrappedMinute(seconds / 60)
        let planned = wrappedMinute(plan.bedtimeMinute + plan.sleepOpportunityMinutes / 2)
        var delta = planned - observed
        if delta > 12 * 60 { delta -= 24 * 60 }
        if delta < -(12 * 60) { delta += 24 * 60 }
        return delta
    }
}
